gui/tiletypes.lua

```diff
@@ -0,0 +1,1667 @@
+            if not self.flat or cursorPos.z == self.first_point.z then
+                df.global.cursor.x = math.max(math.min(cursorPos.x, df.global.world.map.x_count - 1), 0)
+                df.global.cursor.y = math.max(math.min(cursorPos.y, df.global.world.map.y_count - 1), 0)
+                cursorPos = guidm.getCursorPos()
+                self.last_point = cursorPos
+                self:confirm()
+            end
+        elseif dfhack.maps.isValidTilePos(cursorPos.x, cursorPos.y, cursorPos.z) then
+            self.first_point = self.first_point or cursorPos
+        end
+
+        return true
+    end
+
+    local avoid_rect = type(self.avoid_rect) == "function" and self.avoid_rect() or self.avoid_rect
+
+    -- Get the position of the mouse in coordinates local to avoid_rect, if it's specified
+    local mouseFramePos = avoid_rect and self:getMousePos(gui.ViewRect{
+        rect=gui.mkdims_wh(
+            avoid_rect.x1,
+            avoid_rect.y1,
+            avoid_rect.width,
+            avoid_rect.height
+        )
+    })
+
+    if keys._MOUSE_L and not mouseFramePos then
+        -- If left click and the mouse is not in the avoid_rect
+        if self.first_point and not self.last_point then
+            if not self.flat or mousePos.z == self.first_point.z then
+                local inBoundsMouse = {
+                    x = math.max(math.min(mousePos.x, df.global.world.map.x_count - 1), 0),
+                    y = math.max(math.min(mousePos.y, df.global.world.map.y_count - 1), 0),
+                    z = mousePos.z,
+                }
+                self.last_point = inBoundsMouse
+                self:confirm()
+            end
+        elseif dfhack.maps.isValidTilePos(mousePos.x, mousePos.y, mousePos.z) then
+            self.first_point = self.first_point or mousePos
+        end
+
+        return true
+    end
+
+    -- Switch to the cursor if the cursor was moved (Excluding up and down a Z level)
+    local filteredKeys = utils.clone(keys)
+    filteredKeys["CURSOR_DOWN_Z"] = nil
+    filteredKeys["CURSOR_UP_Z"] = nil
+    self.useCursor = self.useCursor or guidm.getMapKey(filteredKeys)
+        or (self.lastCursorPos and (self.lastCursorPos.x ~= cursorPos.x or self.lastCursorPos.y ~= cursorPos.y))
+    self.lastCursorPos = cursorPos
+
+    return false
+end
+
+function BoxSelection:onRenderFrame(dc, rect)
+    -- Switch to cursor if the mouse is offscreen, or if it hasn't moved
+    self.useCursor = (self.useCursor or (self.lastMousePos and (self.lastMousePos.x < 0 or self.lastMousePos.y < 0)))
+        and self.lastMousePos.x == df.global.gps.precise_mouse_x and self.lastMousePos.y == df.global.gps.precise_mouse_y
+    self.lastMousePos = { x= df.global.gps.precise_mouse_x, y= df.global.gps.precise_mouse_y }
+
+    if self.screen and self.screen.subviews.dimensions_tooltip then
+        self.screen.subviews.dimensions_tooltip.visible = not self.useCursor
+    end
```

Comment:
This is because the BoxSelection system was designed around being able to use either keyboard or mouse, but DimensionsTooltip only works when using a mouse. When trying to use it with the keyboard cursor, the dimensions and position of the tooltip are wrong, so I ended up just hiding it.
![image](https://github.com/DFHack/scripts/assets/26845479/7c8437ef-39e7-4ec9-bdc9-81fb503a9a6b)

Reply:
ok, I see. I was thinking that you could just feed the keyboard cursor dimensions to the tooltip, but even if you do, it would appear in the wrong place. That could be rectified in `DimensionsTooltip`, though. it could take a callback for indicating whether it should display the tooltip near the mouse cursor or the keyboard cursor.